Example 2 (Celsius to Fahrenheit) Convert done

F20170604 opened the pull request:
Original File Link : https://colab.research.google.com/github/tensorflow/examples/blob/master/courses/udacity_intro_to_tensorflow_for_deep_learning/l02c01_celsius_to_fahrenheit.ipynb#scrollTo=xrpFFlgYhCty

Converted File Link: https://colab.research.google.com/drive/1Gp5ypYsDZ3QPzgNBCUbW3h6fMj3asz9y

F20170604 committed:
Example 2 (Celsius to Fahrenheit) Convert done

File Converted : https://colab.research.google.com/github/tensorflow/examples/blob/master/courses/udacity_intro_to_tensorflow_for_deep_learning/l02c01_celsius_to_fahrenheit.ipynb#scrollTo=xrpFFlgYhCty

BradLarson commented on S4TF_Tutorial_2.ipynb:
```diff
@@ -0,0 +1,878 @@
+        "colab": {}
+      },
+      "source": [
+        "import TensorFlow\n",
```
This second `import TensorFlow` can probably be removed, because it was already imported above.

F20170604 replied:
Oh yes sure 👍

rxwei commented:
Could you please review the [Swift API Design Guidelines](https://swift.org/documentation/api-design-guidelines) and the [Google Swift Style Guide](https://google.github.io/swift) to make sure your code conform to those guidelines? In particular, if you could spend some time to make sure the name capitalization and horizontal whitespaces follow the guidelines, that will be very helpful!

> Note: The only exception from the Google Swift Style Guide is that we apply 4-space indentation.

Ayush517 commented:
> Could you please review the [Swift API Design Guidelines](https://swift.org/documentation/api-design-guidelines) and the [Google Swift Style Guide](https://google.github.io/swift) to make sure your code conforms to those guidelines? In particular, if you could spend some time to make sure the name capitalization and horizontal whitespaces follow the guidelines, that will be very helpful!
> 
> > Note: The only exception from the Google Swift Style Guide is that we apply 4-space indentation.

Sure

F20170604 committed:
Update S4TF_Tutorial_2.ipynb

Ayush517 commented:
@rxwei I have made the necessary changes. For variables, lowerCamelCase is used and 4-space indentation is used in the entire notebook wherever needed. Can you please verify them?
@dynamicwebpaige Can you please check if the explanation needs any changes for this notebook? :smile:

rxwei commented:
Thanks Ayush! I’ll take a closer look tomorrow or Tuesday.

rxwei commented on S4TF_Tutorial_2.ipynb:
```diff
@@ -0,0 +1,881 @@
+        }
+      },
+      "source": [
+        "let unlabeledDataset:[Float] = [10,20,30,40,50,60,70,80,90,100]\n",
```
Please make sure there’s a whitespace after every colon, comma or other punctuation.

F20170604 replied:
Sure 👍 
Done

F20170604 committed:
Whitespaces added before punctuations

dynamicwebpaige commented:
Thanks for making these changes, @Ayush517! @rxwei, are there any other modifications that Ayush should make before this is merged?

rxwei commented on S4TF_Tutorial_2.ipynb:
```diff
@@ -0,0 +1,875 @@
+        "colab": {}
+      },
+      "source": [
+        "struct c2fConv: Layer {\n",
```
Type names should start with an uppercase letter.

rxwei commented on S4TF_Tutorial_2.ipynb:
```diff
@@ -0,0 +1,875 @@
+      },
+      "source": [
+        "struct c2fConv: Layer {\n",
+        "    \n",
```
Remove extra lines after `{` and before `}`.

rxwei commented on S4TF_Tutorial_2.ipynb:
```diff
@@ -0,0 +1,875 @@
+      },
+      "source": [
+        "let hiddenSize: Int = 4\n",
+        "struct c2fConvExp: Layer {\n",
```
Type names should start with an uppercase letter. 

Also, what does "Exp" mean? The Swift API Design Guidelines actually discourage abbreviations. How about renaming `c2fConv` to `CelsiusToFahrenheit`, and renaming `c2fConvExp` to `CelsiusToFahrenheit2`?

F20170604 replied:
This was a part of an experiment for the learners, so I named it as Exp

rxwei replied:
Understood. However, since the entire section is titled "A little experiment", adding "exp" to each variable feels not adding much value but just redundancy. Also, because Swift generally doesn't encourage the use of abbreviations and because "exp" has many possible meanings, names ending with "Exp" can be very confusing.

rxwei commented on S4TF_Tutorial_2.ipynb:
```diff
@@ -0,0 +1,875 @@
+      },
+      "source": [
+        "let celsiusTemperatures: [Float] = [-40, -10,  0,  8, 15, 22,  38]\n",
+        "let input = Tensor<Float>(shape: TensorShape(7, 1), scalars: celsiusTemperatures)\n",
```
[`TensorShape` conforms to `ExpressibleByArrayLiteral`](https://github.com/tensorflow/swift-apis/blob/c1b18f73be9d67c5b2be80c5dfd044791c0c1d2d/Sources/TensorFlow/Core/TensorShape.swift#L23). You can use `[7, 1]` directly.

```suggestion
        "let input = Tensor<Float>(shape: [7, 1], scalars: celsiusTemperatures)\n",
```

rxwei commented on S4TF_Tutorial_2.ipynb:
```diff
@@ -0,0 +1,875 @@
+        "let celsiusTemperatures: [Float] = [-40, -10,  0,  8, 15, 22,  38]\n",
+        "let input = Tensor<Float>(shape: TensorShape(7, 1), scalars: celsiusTemperatures)\n",
+        "let fahrenheitTemperatures: [Float] = [-40,  14, 32, 46, 59, 72, 100]\n",
+        "let output = Tensor<Float>(shape: TensorShape(7, 1), scalars: fahrenheitTemperatures)\n",
```
```suggestion
        "let output = Tensor<Float>(shape: [7, 1], scalars: fahrenheitTemperatures)\n",
```

rxwei commented on S4TF_Tutorial_2.ipynb:
```diff
@@ -0,0 +1,875 @@
+      "source": [
+        "let celsiusTemperatures: [Float] = [-40, -10,  0,  8, 15, 22,  38]\n",
+        "let input = Tensor<Float>(shape: TensorShape(7, 1), scalars: celsiusTemperatures)\n",
+        "let fahrenheitTemperatures: [Float] = [-40,  14, 32, 46, 59, 72, 100]\n",
```
Remove extra whitespace.
```suggestion
        "let fahrenheitTemperatures: [Float] = [-40, 14, 32, 46, 59, 72, 100]\n",
```

rxwei commented on S4TF_Tutorial_2.ipynb:
```diff
@@ -0,0 +1,875 @@
+        "  \n",
+        "}\n",
+        "\n",
+        "var modelExp = c2fConvExp()\n",
```
```suggestion
        "var model = CelciusToFahrenheit2()\n",
```

rxwei replied:
New names will shadow old ones of the same name. No need to disambiguate with "exp" for any variables here.

rxwei commented on S4TF_Tutorial_2.ipynb:
```diff
@@ -0,0 +1,875 @@
+        "\n",
+        "var modelExp = c2fConvExp()\n",
+        "\n",
+        "let optimizerExp = Adam(for: modelExp, learningRate: 0.1)\n",
```
```suggestion
        "let optimizer = Adam(for: model2, learningRate: 0.1)\n",
```

rxwei commented on S4TF_Tutorial_2.ipynb:
```diff
@@ -0,0 +1,875 @@
+        "        optimizer.update(&model.allDifferentiableVariables, along: grad)\n",
+        "    epochLoss = loss.scalarized()\n",
+        "    trainLossResults.append(epochLoss)\n",
+        "    if epoch % 20 == 0 {\n",
```
```suggestion
        "    if epoch.isMultiple(of: 20) {\n",
```

rxwei commented on S4TF_Tutorial_2.ipynb:
```diff
@@ -0,0 +1,875 @@
+        "        let pred = model(input)\n",
+        "        return meanSquaredError(predicted: pred, expected: output)\n",
+        "    }\n",
+        "        optimizer.update(&model.allDifferentiableVariables, along: grad)\n",
```
Could you correct the indentation here?
```suggestion
        "    optimizer.update(&model.allDifferentiableVariables, along: grad)\n",
```

rxwei commented on S4TF_Tutorial_2.ipynb:
```diff
@@ -0,0 +1,875 @@
+        "    optimizerExp.update(&modelExp.allDifferentiableVariables, along: gradExp)\n",
+        "    epochLossExp = lossExp.scalarized()\n",
+        "    trainLossResultsExp.append(epochLossExp)\n",
+        "    if epoch % 20 == 0 {\n",
```
```suggestion
        "    if epoch.isMultiple(of: 20) {\n",
```

F20170604 committed:
Changes made as proposed made by Richard

Type names now changed and are starting with an uppercase letter.
Abbreviation "Exp" removed entirely from the notebook and replaced with "2"
Indentation corrected

rxwei commented on S4TF_Tutorial_2.ipynb:
```diff
@@ -0,0 +1,877 @@
+        "\n",
+        "for epoch in 1...epochCount {\n",
+        "    var epochLoss2: Float = 0\n",
+        "    let (loss2, grad2) = model2.valueWithGradient { (model2: CelsiusToFahrenheit2) -> Tensor<Float> in\n",
```
```suggestion
        "    let (loss2, grad2) = model2.valueWithGradient { model -> Tensor<Float> in\n",
```

Drop redundant type information.

rxwei commented on S4TF_Tutorial_2.ipynb:
```diff
@@ -0,0 +1,877 @@
+        "\n",
+        "for epoch in 1...epochCount {\n",
+        "    var epochLoss: Float = 0\n",
+        "    let (loss, grad) = model.valueWithGradient { (model: CelsiusToFahrenheit) -> Tensor<Float> in\n",
```
```suggestion
        "    let (loss, grad) = model.valueWithGradient { model -> Tensor<Float> in\n",
```

rxwei reviewed (commented):
In the experiment section, I don't think it's particularly meaningful to suffix any variable with `2` (or `Exp` earlier). They are already in a section titled "experiment". So long as the model type is different (`CelsiusToFahrenheit2`), it's less confusing to the reader to just use variable names like `model`, `loss` and `grad`.

rxwei reviewed (approved):
LGTM after comments above are addressed. Thanks!

F20170604 committed:
Conventional variable names used now

Ayush517 commented:
Can this be merged now? All issues addressed @rxwei

dynamicwebpaige commented:
@Ayush517 @rxwei: SGTM! Thanks to you both for reviewing and resolving the issues. Excited to have this notebook as an example!

rxwei commented:
Looks great. Thanks for your patience in addressing PR feedback!

Ayush517 commented:
> Looks great. Thanks for your patience in addressing PR feedback!

Oh certainly. I loved it. Learnt so so many things during this.
I hope feedbacks like this happen to all the further PRs. :+1:
